modules/replication/module.py

```diff
@@ -27,15 +64,26 @@ def run(self):
 
     def flush_local(self):
         """Resets all local variables."""
-        raise NotImplementedError
+        self.seq_n = 0
+        self.rep = [deepcopy(self.DEF_STATE) for i in range(
+                                                        self.number_of_nodes)]
 
     def msg(self, status, processor_j):
         """Returns requests reported to p_i from processor_j with status."""
-        raise NotImplementedError
+        request_set = set()
+        for request_pair in self.rep[processor_j].get(REQ_Q):
+            if request_pair.get(STATUS) == status:
+                request_set.add(request_pair.get(REQUEST))
+        return request_set
 
     def last_exec(self):
-        """Returns last request sequence number executed."""
-        raise NotImplementedError
+        """Returns last request (highest sequence number) executed."""
```

Comment:
Edit, since requests are always added in order to `r_log`, the last executed request with the highest sequence number can be accessed as `r_log[len(r_log) - 1]` in `O(1)` time 🎉

Reply:
Done!